### PR TITLE
Add endActionState and fromActionState to fetchActions

### PR DESCRIPTION
**Pull request by MartinMinkov**
# Description
:link: Resolves https://github.com/o1-labs/snarkyjs/issues/802

This PR introduces `fromActionState` and `endActionState` parameters to the `fetchActions` function in SnarkyJS. This enhancement allows us to fetch only the necessary actions to compute the latest actions state, instead of retrieving the entire actions history of a zkApp as done previously.

With these improvements, we no longer need to start from an empty action state and apply hashes to derive the current action state by analyzing each block from the beginning of a zkApp's history. Instead, we can utilize the new `actionStateTwo` field returned by the Archive Node API. As `actionStateTwo` represents the second latest action state calculated, it serves as a safe starting point for deriving the most recent action state. We can now apply AccountUpdates to the `actionStateTwo` value to compute the current action state.

Furthermore, users can now provide a `fromActionState` parameter to the Archive Node API, which will return actions only after the specified action state. This feature offers a filtering option in SnarkyJS that was previously unavailable.

Please note that this PR relies on https://github.com/o1-labs/Archive-Node-API/pull/71.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -30,6 +30,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   - The internal event type now includes event data and transaction information as separate objects, allowing for more accurate information about each event and its associated transaction.
 - Removed multiple best tip blocks when fetching action data https://github.com/o1-labs/snarkyjs/pull/817
   - Implemented a temporary fix that filters out multiple best tip blocks, if they exist, while fetching actions. This fix will be removed once the related issue in the Archive-Node-API repository (https://github.com/o1-labs/Archive-Node-API/issues/7) is resolved.
+- New `fromActionHash` and `endActionHash` parameters for fetchActions function in SnarkyJS https://github.com/o1-labs/snarkyjs/pull/828
```
Comment (mitschabaude) [inline CHANGELOG.md:33]: could we rename to `fromActionState` and `endActionState`?

File: src/lib/mina.ts [hunk 2]
```diff
@@ -148,6 +149,11 @@ type DeprecatedFeePayerSpec =
     })
   | undefined;
 
+type ActionHashes = {
```
Comment (mitschabaude) [inline src/lib/mina.ts:152]: rename to `ActionStates`?

File: src/lib/mina.ts [hunk 3]
```diff
@@ -340,6 +346,7 @@ interface Mina {
   ) => ReturnType<typeof Fetch.fetchEvents>;
   getActions: (
     publicKey: PublicKey,
+    actionHashes: ActionHashes,
```
Comment (mitschabaude) [inline src/lib/mina.ts:349]: I would make this argument optional, otherwise users have to pass in an empty object

File: src/lib/mina.ts [hunk 4]
```diff
@@ -581,6 +588,7 @@ function LocalBlockchain({
     },
     getActions(
       publicKey: PublicKey,
+      _actionHashes: ActionHashes,
```
Comment (mitschabaude) [inline src/lib/mina.ts:591]: I'm afraid that ignoring the `_actionHashes` input here leads to inconsistent behavior between the `Mina.Network` and `Mina.LocalBlockchain` interfaces. Maybe you could move the logic in `SmartContract.getActions` which filters the actions here instead?

**Review by mitschabaude — APPROVED** [review]
Mostly looks good, although I would change the 'actionHash' terminology to 'actionState', and make sure the local and remote Mina interfaces are consistent.

A general comment is that we don't have good test coverage for `getActions`. Ideally, there'd be a unit test which confirms expected behavior on LocalBlockchain (for cases like multiple actions in an account update / in a block / computing action states from multiple blocks with actions). Plus a manual test which confirms it with an archive node, for the time being where the Berkeley test doesn't work in CI

File: src/lib/mina.ts (hunk 4 above)
Comment (MartinMinkov) [reply src/lib/mina.ts]: Moved the logic over! :D

File: CHANGELOG.md [hunk 5]
```diff
@@ -30,6 +30,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   - The internal event type now includes event data and transaction information as separate objects, allowing for more accurate information about each event and its associated transaction.
 - Removed multiple best tip blocks when fetching action data https://github.com/o1-labs/snarkyjs/pull/817
   - Implemented a temporary fix that filters out multiple best tip blocks, if they exist, while fetching actions. This fix will be removed once the related issue in the Archive-Node-API repository (https://github.com/o1-labs/Archive-Node-API/issues/7) is resolved.
+- New `fromActionState` and `endActionState` parameters for fetchActions function in SnarkyJS https://github.com/o1-labs/snarkyjs/pull/828
```
Comment (mitschabaude) [inline CHANGELOG.md:33]: could we also, under the "breaking changes" section, mention the API change in `getActions` from `fromActionHash` etc to `fromActionState`?

**Pull request merged**